### PR TITLE
gh-112346: Always set OS byte to 255, simpler gzip.compress function.

### PR DESCRIPTION
**Pull request by rhpvorderman**
* Issue: gh-112346

----
📚 Documentation preview 📚: https://cpython-previews--120486.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/gzip.rst [hunk 1]
```diff
@@ -188,17 +188,16 @@ The module defines the following items:
...
-   the :class:`GzipFile` constructor above. When *mtime* is set to ``0``, this
-   function is equivalent to :func:`zlib.compress` with *wbits* set to ``31``.
-   The zlib function is faster.
+   the :class:`GzipFile` constructor above.
 
    .. versionadded:: 3.2
    .. versionchanged:: 3.8
       Added the *mtime* parameter for reproducible output.
    .. versionchanged:: 3.11
-      Speed is improved by compressing all data at once instead of in a
```
Comment (serhiy-storchaka) [inline Doc/library/gzip.rst]: Something is missed.

File: Doc/library/gzip.rst [hunk 2]
```diff
@@ -188,17 +188,16 @@ The module defines the following items:
...
       streamed fashion. Calls with *mtime* set to ``0`` are delegated to
       :func:`zlib.compress` for better speed.
+   .. versionchanged:: 3.14
```
Comment (serhiy-storchaka) [inline Doc/library/gzip.rst]: Please add an empty line before this directive.

File: Lib/gzip.py [hunk 3]
```diff
@@ -580,43 +580,20 @@ def _rewind(self):
...
-    if mtime == 0:
-        # Use zlib as it creates the header with 0 mtime by default.
-        # This is faster and with less overhead.
-        return zlib.compress(data, level=compresslevel, wbits=31)
-    header = _create_simple_gzip_header(compresslevel, mtime)
-    trailer = struct.pack("<LL", zlib.crc32(data), (len(data) & 0xffffffff))
-    # Wbits=-15 creates a raw deflate block.
-    return (header + zlib.compress(data, level=compresslevel, wbits=-15) +
-            trailer)
+    gzip_data = zlib.compress(data, level=compresslevel, wbits=31)
```
Comment (serhiy-storchaka) [inline Lib/gzip.py:591]: Please comment why `wbits=31` is used.

File: Lib/test/test_gzip.py [hunk 4]
```diff
@@ -714,14 +714,25 @@ def test_compress_mtime(self):
                         self.assertEqual(f.mtime, mtime)
 
     def test_compress_correct_level(self):
-        # gzip.compress calls with mtime == 0 take a different code path.
         for mtime in (0, 42):
             with self.subTest(mtime=mtime):
                 nocompress = gzip.compress(data1, compresslevel=0, mtime=mtime)
-                yescompress = gzip.compress(data1, compresslevel=1, mtime=mtime)
+                yescompress = gzip.compress(data1, compresslevel=1,
+                                                 mtime=mtime)
```
Comment (serhiy-storchaka) [inline Lib/test/test_gzip.py]: This looks like a not related change.

File: Misc/NEWS.d/next/Library/2024-06-12-10-00-31.gh-issue-90425.5CfkKG.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+The OS byte in gzip headers is now always set to 255 when using
+gzip.compress.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2024-06-12-10-00-31.gh-issue-90425.5CfkKG.rst]:
```suggestion
:func:`gzip.compress`.
```

**Comment by rhpvorderman** [comment]
Thanks for the comments @serhiy-storchaka . They have been addressed.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

File: Lib/gzip.py [hunk 6]
```diff
@@ -580,43 +580,21 @@ def _rewind(self):
...
-    return (header + zlib.compress(data, level=compresslevel, wbits=-15) +
-            trailer)
+    # Wbits=31 automatically includes a gzip header and trailer.
+    gzip_data = zlib.compress(data, level=compresslevel, wbits=31)
+    if mtime is None:
+        mtime = time.time()
+    # Reuse gzip header created by zlib, replace mtime and OS byte for
+    # consistency.
+    header = struct.pack("<4sLBB", gzip_data, int(mtime), gzip_data[8], 255)
+    return header + gzip_data[10:]
```
Comment (gpshead) [inline Lib/gzip.py:597]: This causes a large data copy and requires double the memory.  We really should do this header rewriting on the original bytes buffer in C within zlib.compress, otherwise it is a performance regression.

Comment (gpshead) [reply Lib/gzip.py:597]: Regression is the wrong word, the old code this replaces did an equivalent copy... we could just do better.  Given the old code's behavior, don't let this comment block the PR. :)

Comment (serhiy-storchaka) [reply Lib/gzip.py:597]: The old code did two copies.

Also, I suppose that the cost of copying is insignificant in comparison with the cost of compressing. It may be not worth to spent our time on it.

Comment (serhiy-storchaka) [reply Lib/gzip.py:597]: The new code also does two copies, so it is the same.

Comment (rhpvorderman) [reply Lib/gzip.py:597]: The cost of copying is insignificant to the cost of compression. Using zlib.compress, which writes one buffer, is barely faster. At most about 1-2% or so. This is not worth optimizing over IMO.

Comment (rhpvorderman) [reply Lib/gzip.py:597]: > We really should do this header rewriting on the original bytes buffer in C within zlib.compress, otherwise it is a performance regression.

I did one suggestion earlier about improving gzip earlier, but we decided not to go through because of PEP-399 (https://github.com/python/cpython/issues/110283#issuecomment-1910354820). That states that an equivalent python option must always remain present. So moving it to C requires then to have two equivalent code paths to maintain. That is not really worth the 1% performance upgrade.

Comment (gpshead) [reply Lib/gzip.py:597]: The cost is unnecessary double memory use.  /But did you see where I said don't worry about this?

The optimization **is** worthwhile.  But I don't want it done in this PR, if it happens at all, it'll probably be a followup from me. :)  --  PEP-399 does not apply because zlib is not implemented in Python and never will be. The header bytes modification would be done in the `zlib.compress` extension code.

**Review by gpshead — APPROVED** [review]

File: Lib/gzip.py (hunk 6 above)
Comment (rhpvorderman) [reply Lib/gzip.py:597]: I was not aware that PEP-399 did not apply.
As a sidenote, I think more optimally estimating initial memory use using the deflatebound functions is also important as PyMem_Realloc also does a new allocation plus a copy in some occassions, which causes unecessary memory usage.
